### ISSUE TITLE
Equivalent forms of the axiom of replacement

### ISSUE THREAD

**Issue by jkingdon** [metamath/set.mm#4394]
The comment for https://us.metamath.org/mpeuni/f1dmex.html says "This theorem is equivalent to the Axiom of Replacement". Is there is a way to formalize this statement?

First of all, we'd want, I suppose, a notation `REP` analogous to `CHOICE` ( https://us.metamath.org/mpeuni/df-ac.html ) or `EXMID` ( https://us.metamath.org/ileuni/df-exmid.html ).

But how to state the theorem in such a way that it would imply `REP`?

`f1dmex` is `(𝐹:𝐴–1-1→𝐵 ∧ 𝐵 ∈ 𝐶) → 𝐴 ∈ V` . The class `B` isn't much of a problem, it could be replaced by a universally quantified set variable which is similar to what we do in other similar circumstances. But what about ` A `? It has to be a class which is not a priori constrained to be a set, as that's the whole point. Perhaps we can refer to ``( `' F " B )`` (not that I looked carefully at whether replacement would be needed to make that notation express what we want it to express)? Perhaps there is some other solution?

Inspired by the Mastodon thread at https://mathstodon.xyz/@highergeometer/113480205427072831 which isn't about the exact same ax-rep equivalence as `f1dmex` but is about one which is pretty similar.

**Comment by jkingdon** [comment]
~~I think maybe I have a statement of it.~~ _This isn't correct; see below_

```
f1dmexr.1 $e |- ( A. f A. b ( f : dom f -1-1-> b ) -> dom f e. _V ) $.
f1dmexr $p |- |- ( A. w E. y A. z ( A. y ph -> z = y ) -> E. y A. z ( z e. y
      <-> E. w ( w e. x /\ A. y ph ) ) ) $= ... $.
$( $j usage 'f1dmexr' avoids 'ax-rep' ; $)
```

(this is without trying to define a `REP` notation; the version with that notation would just replace the conclusion with `REP`).

**Comment by benjub** [comment]
To formalize equivalences, we do not have to introduce new notation. For instance, we proved several equivalents, labeled `axrep?`, without new notation.

Your `f1dmexr.1` does not formalize REP: it is true without REP, since `f` is a set function.

I think proving `axrep` from all the other axioms plus `f1dmex` would qualify as a proof that `f1dmex` is equivalent to ax-rep.

**Comment by jkingdon** [comment]
> Your `f1dmexr.1` does not formalize REP: it is true without REP, since `f` is a set function.

Oh, you're right. This attempt accidentally excludes the possibility of F as a proper class, so it doesn't work.

> I think proving `axrep` from all the other axioms plus `f1dmex` would qualify as a proof that `f1dmex` is equivalent to ax-rep.

Hmm, like this?

```
f1dmexr2.1 $e |- ( ( F : A -1-1-> B /\ B e. C ) -> A e. _V ) $.
f1dmexr2 $p |- |- ( A. w E. y A. z ( A. y ph -> z = y ) -> E. y A. z ( z e. y
      <-> E. w ( w e. x /\ A. y ph ) ) ) $= ... $.
$( $j usage 'f1dmexr2' avoids 'ax-rep' ; $)
```

This doesn't seem to provide a fully flexible `f1dmex`.  That is, you could satisfy the hypothesis by simply substituting in `(/)` for `A`.

Or maybe you'd need

```
f1dmexr3ax $a |- ( ( F : A -1-1-> B /\ B e. C ) -> A e. _V ) $.
f1dmexr3 $p |- |- ( A. w E. y A. z ( A. y ph -> z = y ) -> E. y A. z ( z e. y
      <-> E. w ( w e. x /\ A. y ph ) ) ) $= ... $.
$( $j usage 'f1dmexr3' avoids 'ax-rep' ; $)
```

which I guess might work but I can't find a lot of examples of this technique so I'm not sure what to think.

**Comment by sctfn** [comment]
I don't think you can state a single formula that is equivalent to REP in ZFC. If you could, then you could finitely axiomatize ZFC (by just stating REP as an axiom). We know that's not possible (that's the whole reason for NBG), so I don't think it's doable.

**Comment by benjub** [comment]
Sorry, I was not clear. What I meant is much more low tech: I think we should be content with a theorem and comment like
```
($ Prove ~ax-rep from ~f1dmex and earlier axioms.  Note that the only use of ~ ax-rep is via f1dmex .$)
axrep9 $p |- ...
```
which we already do in a few instances in set.mm.

@sctfn is right that you cannot in general have such a REP (although beware that we have the axiom scheme of separation ax-sep, so your argument is correct if we do not use ax-sep either (it's probably true if you allow it too, but with a more involved argument)).

In this specific case, however, if your proof uses only a known finite number of instances of the scheme `f1dmex`, then you can put these specific instances as $e-hypotheses.

**Comment by jkingdon** [comment]
> if your proof uses only a known finite number of instances of the scheme `f1dmex`, then you can put these specific instances as $e-hypotheses.

Seems like that's what we'd need to do.

**Comment by LegionMammal978** [comment]
Suppose we want to prove that axiom A is equivalent to theorem T. The general pattern I've seen is to simply rederive A as a theorem (with no extra hypotheses), and expect the reader to verify that A is not used in any step of that theorem except for uses of T itself. But if we want to make it extra clear, we can have a lemma where A is unused and the T instances are hypotheses, then have the final theorem simply insert the T instances into those hypotheses.

Though in this particular case, I'm not sure if it's even true that ax-rep can be derived from [f1dmex](https://us.metamath.org/mpeuni/f1dmex.html). Suppose we want to derive [axrep6](https://us.metamath.org/mpeuni/axrep6.html) `|- ( A. w E* z ph -> E. y A. z ( z e. y <-> E. w e. x ph ) )` with `$d ph y $.`. Let `R = { <. w , z >. | ( w e. x /\ ph ) }`. Then our instance of [f1dmex](https://us.metamath.org/mpeuni/f1dmex.html) will effectively have to be of form `|- ( ( F : ran R -1-1-> x /\ x e. _V ) -> ran R e. _V ) `.

But can we prove the existence of a function `F : ran R --> x` with the appropriate properties? For any `z e. ran R`, we must select a single ``w e. ( `' R " { z } )``. But in general, collecting all of these selected `w`s into a single function `F` can't be done without Choice, since ``( `' R " { z } )`` might have arbitrarily many elements. I'm not sure if our ax-ac/ax-ac2 are usable without Replacement, but if they are, then it's only by coincidence of how they're written.

**Comment by icecream17** [comment]
Consider a function `f = (x in A ↦ f(x))`. Given `A in _V` we want to show the range exists by using f1dmex instead of funimaexg

Take the function `g = (x in A ↦ <x, f(x)>)` and `h = (y in ran g ↦ 1st(y))`. The composition `h o. g` is just the identity function, so the range of `h` = `A`. So the range of `h` exists.

And, `h` is injective: `h(y) = h(z) -> y = z`. This is true because `y` and `z` are in the range of `g`, so `y = g(a), z = g(b)` for some `a, b in A`, so we have `h(g(a)) = h(g(b)) -> a = b -> g(a) = g(b)`

Thus, the hypotheses of f1dmex is satisfied: `h` is injective and the range of `h` exists.
Using f1dmex we get that the domain of `h` = the range of `g` exists. Finally, the range of `g` is `{<x, y>|x in A and y = f(x)}`, or in other words `f` itself. So `f in _V`. And using [rnex](https://us.metamath.org/mpeuni/rnex.html) we get `ran f in _V`. QED

([Derived from ChatGPT...](https://github.com/icecream17/Stuff/blob/main/writing/screencapture-chatgpt-2025-10-20-18_35_03.png))

**Comment by LegionMammal978** [comment]
> And using [rnex](https://us.metamath.org/mpeuni/rnex.html) we get `ran f in _V`.

Ah, that's the trick. I thought [rnex](https://us.metamath.org/mpeuni/rnex.html) and similar theorems would need Replacement to construct the resulting set, but really they just need Union to pull out the elements of the ordered pairs, and it's flexible enough that any sane ordered-pair definition should work. And so when constructing our `F`, we're allowed to tag the elements of the range with the elements of the domain, and avoid needing Choice.
